File: Python/design-a-number-container-system.py
# Time:  ctor:   O(1)
#        change: O(logn)
#        find:   O(1)
# Space: O(n)

from sortedcontainers import SortedList


# sorted list
class NumberContainers(object):

    def __init__(self):
        self.__idx_to_num = {}
        self.__num_to_idxs = collections.defaultdict(SortedList)

    def change(self, index, number):
        """
        :type index: int
        :type number: int
        :rtype: None
        """
        if index in self.__idx_to_num:
            self.__num_to_idxs[self.__idx_to_num[index]].remove(index)
            if not self.__num_to_idxs[self.__idx_to_num[index]]:
                del self.__num_to_idxs[self.__idx_to_num[index]]
        self.__idx_to_num[index] = number
        self.__num_to_idxs[number].add(index)

    def find(self, number):
        """
        :type number: int
        :rtype: int
        """
        return self.__num_to_idxs[number][0] if number in self.__num_to_idxs else -1

File: C++/design-a-number-container-system.cpp
// Time:  ctor:   O(1)
//        change: O(logn)
//        find:   O(1)
// Space: O(n)

// bst
class NumberContainers {
public:
    NumberContainers() {
        
    }
    
    void change(int index, int number) {
        if (idx_to_num_.count(index)) {
            num_to_idxs_[idx_to_num_[index]].erase(index);
            if (empty(num_to_idxs_[idx_to_num_[index]])) {
                num_to_idxs_.erase(idx_to_num_[index]);
            }
        }
        idx_to_num_[index] = number;
        num_to_idxs_[number].emplace(index);
    }
    
    int find(int number) {
        if (num_to_idxs_.count(number)) {
            return *cbegin(num_to_idxs_[number]);
        }
        return -1;
    }

private:
    unordered_map<int, int> idx_to_num_;
    unordered_map<int, set<int>> num_to_idxs_;
};
